pyteal/ast/abi/bool_test.py

```diff
@@ -0,0 +1,332 @@
+    boolAwareStaticByteLength,
+    consecutiveBools,
+    boolSequenceLength,
+    encodeBoolSequence,
+)
+
+# this is not necessary but mypy complains if it's not included
+from ... import CompileOptions
+
+options = CompileOptions(version=5)
+
+
+def test_Bool_str():
+    boolType = abi.Bool()
+    assert str(boolType) == "bool"
+
+
+def test_Bool_is_dynamic():
+    boolType = abi.Bool()
+    assert not boolType.is_dynamic()
+
+
+def test_Bool_has_same_type_as():
+    boolType = abi.Bool()
+    assert boolType.has_same_type_as(abi.Bool())
+
+    for otherType in (
+        abi.Byte(),
+        abi.Uint64(),
+        abi.StaticArray(boolType, 1),
+        abi.DynamicArray(boolType),
+    ):
+        assert not boolType.has_same_type_as(otherType)
+
+
+def test_Bool_new_instance():
+    boolType = abi.Bool()
+    assert type(boolType.new_instance()) is abi.Bool
+
+
+def test_Bool_set_static():
+    boolType = abi.Bool()
+    for value in (True, False):
+        expr = boolType.set(value)
+        assert expr.type_of() == TealType.none
+        assert not expr.has_return()
+
+        expected = TealSimpleBlock(
+            [
+                TealOp(None, Op.int, 1 if value else 0),
+                TealOp(None, Op.store, boolType.stored_value.slot),
+            ]
+        )
+
+        actual, _ = expr.__teal__(options)
+        actual.addIncoming()
+        actual = TealBlock.NormalizeBlocks(actual)
+
+        with TealComponent.Context.ignoreExprEquality():
+            assert actual == expected
+
+
+def test_Bool_set_expr():
+    boolType = abi.Bool()
+    expr = boolType.set(Int(0).Or(Int(1)))
+    assert expr.type_of() == TealType.none
+    assert not expr.has_return()
+
+    expected = TealSimpleBlock(
+        [
+            TealOp(None, Op.int, 0),
+            TealOp(None, Op.int, 1),
+            TealOp(None, Op.logic_or),
+            TealOp(None, Op.store, boolType.stored_value.slot),
+        ]
+    )
+
+    actual, _ = expr.__teal__(options)
+    actual.addIncoming()
+    actual = TealBlock.NormalizeBlocks(actual)
+
+    with TealComponent.Context.ignoreExprEquality():
+        assert actual == expected
+
+
+def test_Bool_set_copy():
+    other = abi.Bool()
+    boolType = abi.Bool()
+    expr = boolType.set(other)
+    assert expr.type_of() == TealType.none
+    assert not expr.has_return()
+
+    expected = TealSimpleBlock(
+        [
+            TealOp(None, Op.load, other.stored_value.slot),
+            TealOp(None, Op.store, boolType.stored_value.slot),
+        ]
+    )
+
+    actual, _ = expr.__teal__(options)
+    actual.addIncoming()
+    actual = TealBlock.NormalizeBlocks(actual)
+
+    with TealComponent.Context.ignoreExprEquality():
+        assert actual == expected
+
+
+def test_Bool_get():
+    boolType = abi.Bool()
+    expr = boolType.get()
+    assert expr.type_of() == TealType.uint64
+    assert not expr.has_return()
+
+    expected = TealSimpleBlock([TealOp(expr, Op.load, boolType.stored_value.slot)])
```

Comment:
just one dumb question from my side: what is the difference for `TealOp` to have first argument to have either `Expr` or `None`? Or so to say, in what scenario `TealOp` will want to accept `Expr`, and otherwise when do we want to give `None`?

Reply:
The first argument is the `Expr` which created the `TealOp`, used only for error messages. For testing, if it's easy to construct the expected response which correct `Expr`s, then generally I think we should do that. Otherwise, None should be used for the `Expr` and the context `TealComponent.Context.ignoreExprEquality()` should be enabled before comparing with `==`